### PR TITLE
Torch compile graph fix

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Thanks to @Chillee

I'm testing some possible improvements to `torch.compile`. Currently running the following script for **Stable Diffusion**:

```py
from diffusers import StableDiffusionPipeline
import torch

path = "runwayml/stable-diffusion-v1-5"

run_compile = True  # Set True / False

pipe = StableDiffusionPipeline.from_pretrained(path, torch_dtype=torch.float16)
pipe = pipe.to("cuda:0")
pipe.unet.to(memory_format=torch.channels_last)

if run_compile:
    print("Run torch compile")
    pipe.unet = torch.compile(pipe.unet, mode="reduce-overhead", fullgraph=True)

prompt = "ghibli style, a fantasy landscape with castles"

for _ in range(3):
    images = pipe(prompt=prompt).images
```

and the following for **IFPipeline**:

```py
from diffusers import DiffusionPipeline
from diffusers.utils import pt_to_pil
import torch

# stage 1
pipe = DiffusionPipeline.from_pretrained("DeepFloyd/IF-I-M-v1.0", variant="fp16", text_encoder=None, torch_dtype=torch.float16).to("cuda")

pipe.unet.to(memory_format=torch.channels_last)
pipe.unet = torch.compile(pipe.unet, mode="reduce-overhead", fullgraph=True)

prompt = "the blue hulk"

prompt_embeds = torch.randn((1, 2, 4096), dtype=torch.float16)
neg_prompt_embeds = torch.randn((1, 2, 4096), dtype=torch.float16)

image = pipe(prompt_embeds=prompt_embeds, negative_prompt_embeds=neg_prompt_embeds).images
```

### CODE REVIEW

File: src/diffusers/utils/__init__.py [hunk 1]
```diff
@@ -81,7 +81,7 @@
 from .logging import get_logger
 from .outputs import BaseOutput
 from .pil_utils import PIL_INTERPOLATION, numpy_to_pil, pt_to_pil
-from .torch_utils import is_compiled_module, randn_tensor
+from .torch_utils import is_compiled_module, maybe_allow_in_graph, randn_tensor
```
Comment (patrickvonplaten) [inline src/diffusers/utils/__init__.py:84]:
```suggestion
from .torch_utils import is_compiled_module, randn_tensor
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 2]
```diff
@@ -868,15 +869,16 @@ def custom_forward(*inputs):
                     hidden_states,
                     encoder_hidden_states=encoder_hidden_states,
                     cross_attention_kwargs=cross_attention_kwargs,
-                ).sample
+                    return_dict=False,
+                )[0]
 
-            output_states += (hidden_states,)
+            output_states = output_states + (hidden_states,)
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:875]: I had to do these changes as well to get torch compile to work

**Comment by patrickvonplaten** [comment]
Listing some first results here:

On RTX4090 GPU and the following setup:

```
- `diffusers` version: 0.17.0.dev0
- Platform: Linux-5.15.0-69-generic-x86_64-with-glibc2.35
- Python version: 3.10.6
- PyTorch version (GPU?): 2.0.0+cu117 (True)
- Huggingface_hub version: 0.13.3
- Transformers version: 4.28.0.dev0
- Accelerate version: 0.18.0
- xFormers version: not installed
```

I'm getting for **Stable Diffusion**:
```
eager: 11.95it/s
torch.compile: 13.63it/s
```

**Comment by patrickvonplaten** [comment]
Getting a much bigger speed-up on A100s with CUDA 11.4, for **Stable Diffusion**:

```py
eager: 22.30it/s
torch.compile: 42.62it/s
```

**Comment by patrickvonplaten** [comment]
Tested code again with PyTorch nightly and now it also works for IF. Currently don't have access to a A100 so will post results later

**Comment by mvpatel2000** [comment]
@patrickvonplaten were you able to get any updated numbers on this? very curious about results

**Comment by Chillee** [comment]
@mvpatel2000 Updated benchmarks are here: https://github.com/huggingface/diffusers/pull/3313#issuecomment-1531775103

And yes, the baseline is already using F.sdpa (i.e. flash-attention/xformers mem_efficient_attention).

**Pull request merged**